// @commerce-apps/raml-toolkit/lib/index.js
"use strict";
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var bearerToken_1 = require("./download/bearerToken");
exports.getBearer = bearerToken_1.getBearer;
var logger_1 = require("./common/logger");
exports.ramlToolLogger = logger_1.ramlToolLogger;
const amf = tslib_1.__importStar(require("amf-client-js"));
exports.amf = amf;
var amf_client_js_1 = require("amf-client-js");
exports.model = amf_client_js_1.model;
tslib_1.__exportStar(require("./generate"), exports);
var parser_1 = require("./common/parser");
exports.getAllDataTypes = parser_1.getAllDataTypes;
exports.getApiName = parser_1.getApiName;
exports.getNormalizedName = parser_1.getNormalizedName;
exports.parseRamlFile = parser_1.parseRamlFile;
exports.resolveApiModel = parser_1.resolveApiModel;
