// commerce-sdk/dist/customer/shopperLogin/shopperLogin.js
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShopperLogin = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@commerce-apps/core");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
/**
* [Shopper Login](https://developer.commercecloud.com/s/api-details/)
* ==================================
*
* *Enable shoppers to login via federation to any external IDP, get access token and use it to access shopper APIs*<br />
*
* Simple example:
*
* ```typescript
*   import { Customer, ClientConfig } from "commerce-sdk";
*   // or
*   const { Customer, ClientConfig } = require("commerce-sdk");
*
*   const clientConfig: ClientConfig = { ... };
*   const shopperLoginClient = new Customer.ShopperLogin(clientConfig);
* ```
*
* Example with shopper auth
* ```typescript
* const clientConfig: ClientConfig = {
*   parameters: {
*     clientId: "XXXXXX",
*     organizationId: "XXXX",
*     shortCode: "XXX",
*     siteId: "XX"
*   }
* };
* token = await helpers.getShopperToken(clientConfig, { type: "guest" });
* clientConfig.headers["authorization"] = token.getBearerHeader();
* const shopperLoginClient = new Customer.ShopperLogin(clientConfig);
* ```
*
* <span style="font-size:.7em; display:block; text-align: right">
* API Version: 1.1.0<br />
* Last Updated: 2020-07-17T19:05:14.764Z<br />
* </span>
* @beta
*
*

*/
class ShopperLogin extends core_1.BaseClient {
  /**
   *
   */
  constructor(config) {
    super(config);
    if (!!!config.baseUri) {
      this.clientConfig.baseUri =
        "https://{shortCode}.api.commercecloud.salesforce.com/shopper/auth/{version}";
    }
  }
  /**
   * Get authorization code after authenticating a user against an IDP. This is the first step of the Authorization code grant flow, where a user is able to login via federation to the IDP configured for the client, and gets an authorization code after succesful login via a callback/redirect URI. Mandatory fields are the shopping app's clientId, redirectURI, responseType and authType. This API can be called from the front channel (from the browser).
   *
   * @param options - An object containing the options for this method.
   * @param parameters - An object containing the parameters for this method.
   * @param organizationId -
   * @param redirectUri - The URL to which the server redirects the browser after the user grants the authorization.  The URI must be pre-registered.
   * @param responseType - Must be 'code'.  'code' indicates that the client wants an authorization code (authorization code grant flow).
   * @param clientId - The client id obtained during application registration.
   * @param scope - (Not Supported)
   * @param state - Value to be sent by the client to determine the state between the authorization request and the server response.  Optional but strongly recommended.
   * @param uuid - Id that uniquely identifies the shopper. This is an optional field, a new UUID will be generated if this field is empty.
   * @param hint - Idp name that can be optionally added to redirect directly to skipping the Idp selection step.
   * @param headers - An object literal of key value pairs of the headers to be
   * sent with this request.
   * @param rawResponse - Set to true to return entire Response object instead of DTO.
   *
   * @returns A promise of type Response if rawResponse is true, a promise of type void otherwise.
   *
   * @beta
   *
   */
  authorizeCustomer(options, rawResponse) {
    const parameters = options && options.parameters ? options.parameters : {};
    const pathParameters = {
      organizationId:
        parameters["organizationId"] !== undefined
          ? parameters["organizationId"]
          : this.clientConfig.parameters["organizationId"],
    };
    const queryParameters = lodash_1.default.omit(
      parameters,
      lodash_1.default.keys(pathParameters)
    );
    queryParameters["clientId"] =
      queryParameters["clientId"] !== undefined
        ? parameters["clientId"]
        : this.clientConfig.parameters["clientId"];
    // @ts-ignore
    return core_1.StaticClient.get({
      client: this,
      rawResponse: rawResponse,
      path: "/organizations/{organizationId}/oauth2/authorize",
      pathParameters: pathParameters,
      queryParameters: queryParameters,
      retrySettings: (options || {}).retrySettings,
      headers: (options || {}).headers,
    });
  }
  /**
   * Get the shopper or guest JWT/access token along with a refresh token. This is the second step of the OAuth2 Authorization code flow where a client app is able to get an access token for the shopper through the backchannel (a trusted server side) by passing in the client credentials and the authorization code retrieved from authorize endpoint. Mandatory fields when using the token endpoint to get the accesstoken are clientId, clientSecret, authorizationCode, grantType of authorization_code. \<br/\>\<br/\> As a guest user get the shopper JWT / access token along with a refresh token. This is where a client app is able to get an access token for the guest user through the back-channel (a trusted server side) by passing in the client credentials. Mandatory fields when asking for the guest token endpoint to get the access token are clientId, clientSecret, grantType of client_credentials.\<br/\>\<br/\> Mandatory fields when using using the token endpoint to refresh an access token are refreshToken, grantType of refresh_token.
   *
   * @param options - An object containing the options for this method.
   * @param parameters - An object containing the parameters for this method.
   * @param organizationId -
   * @param headers - An object literal of key value pairs of the headers to be
   * sent with this request.
   * @param body - The data to send as the request body.
   * @param rawResponse - Set to true to return entire Response object instead of DTO.
   *
   * @returns A promise of type Response if rawResponse is true, a promise of type ShopperLogin.TokenResponse otherwise.
   *
   * @beta
   *
   */
  getAccessToken(options, rawResponse) {
    const parameters = options && options.parameters ? options.parameters : {};
    const pathParameters = {
      organizationId:
        parameters["organizationId"] !== undefined
          ? parameters["organizationId"]
          : this.clientConfig.parameters["organizationId"],
    };
    const queryParameters = lodash_1.default.omit(
      parameters,
      lodash_1.default.keys(pathParameters)
    );
    // @ts-ignore
    return core_1.StaticClient.post({
      client: this,
      rawResponse: rawResponse,
      path: "/organizations/{organizationId}/oauth2/token",
      pathParameters: pathParameters,
      queryParameters: queryParameters,
      retrySettings: (options || {}).retrySettings,
      headers: (options || {}).headers,
      body: (options || {}).body,
    });
  }
  /**
   * Invalidate the refresh token. A basic auth header with base64 encoded clientId:secret is required in the Authorization header and the refresh token to be revoked is required in the body.
   *
   * @param options - An object containing the options for this method.
   * @param parameters - An object containing the parameters for this method.
   * @param organizationId -
   * @param headers - An object literal of key value pairs of the headers to be
   * sent with this request.
   * @param body - The data to send as the request body.
   * @param rawResponse - Set to true to return entire Response object instead of DTO.
   *
   * @returns A promise of type Response if rawResponse is true, a promise of type Object otherwise.
   *
   * @beta
   *
   */
  revokeToken(options, rawResponse) {
    const parameters = options && options.parameters ? options.parameters : {};
    const pathParameters = {
      organizationId:
        parameters["organizationId"] !== undefined
          ? parameters["organizationId"]
          : this.clientConfig.parameters["organizationId"],
    };
    const queryParameters = lodash_1.default.omit(
      parameters,
      lodash_1.default.keys(pathParameters)
    );
    // @ts-ignore
    return core_1.StaticClient.post({
      client: this,
      rawResponse: rawResponse,
      path: "/organizations/{organizationId}/oauth2/revoke",
      pathParameters: pathParameters,
      queryParameters: queryParameters,
      retrySettings: (options || {}).retrySettings,
      headers: (options || {}).headers,
      body: (options || {}).body,
    });
  }
}
exports.ShopperLogin = ShopperLogin;
