// @commerce-apps/raml-toolkit/lib/download/bearerToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
const MULESOFT_LOGIN = "https://anypoint.mulesoft.com/accounts/login";
function getBearer(username, password) {
    return node_fetch_1.default(MULESOFT_LOGIN, {
        method: "post",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({
            username: username,
            password: password
        })
    }).then((response) => {
        if (response.ok) {
            return response.json().then(json => json["access_token"]);
        }
        switch (response.status) {
            case 401: {
                throw new Error("Invalid username/password");
            }
            default: {
                throw new Error(`Unknown Error ${response.status}: ${response.statusText}`);
            }
        }
    });
}
exports.getBearer = getBearer;
