// commerce-sdk/dist/pricing/pricing.js
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShopperGiftCertificates = exports.ShopperPromotions = exports.Assignments = exports.Campaigns = exports.Coupons = exports.GiftCertificates = exports.SourceCodeGroups = exports.Promotions = void 0;
const promotions_1 = require("./promotions/promotions");
Object.defineProperty(exports, "Promotions", {
  enumerable: true,
  get: function () {
    return promotions_1.Promotions;
  },
});
const sourceCodeGroups_1 = require("./sourceCodeGroups/sourceCodeGroups");
Object.defineProperty(exports, "SourceCodeGroups", {
  enumerable: true,
  get: function () {
    return sourceCodeGroups_1.SourceCodeGroups;
  },
});
const giftCertificates_1 = require("./giftCertificates/giftCertificates");
Object.defineProperty(exports, "GiftCertificates", {
  enumerable: true,
  get: function () {
    return giftCertificates_1.GiftCertificates;
  },
});
const coupons_1 = require("./coupons/coupons");
Object.defineProperty(exports, "Coupons", {
  enumerable: true,
  get: function () {
    return coupons_1.Coupons;
  },
});
const campaigns_1 = require("./campaigns/campaigns");
Object.defineProperty(exports, "Campaigns", {
  enumerable: true,
  get: function () {
    return campaigns_1.Campaigns;
  },
});
const assignments_1 = require("./assignments/assignments");
Object.defineProperty(exports, "Assignments", {
  enumerable: true,
  get: function () {
    return assignments_1.Assignments;
  },
});
const shopperPromotions_1 = require("./shopperPromotions/shopperPromotions");
Object.defineProperty(exports, "ShopperPromotions", {
  enumerable: true,
  get: function () {
    return shopperPromotions_1.ShopperPromotions;
  },
});
const shopperGiftCertificates_1 = require("./shopperGiftCertificates/shopperGiftCertificates");
Object.defineProperty(exports, "ShopperGiftCertificates", {
  enumerable: true,
  get: function () {
    return shopperGiftCertificates_1.ShopperGiftCertificates;
  },
});
