// @commerce-apps/raml-toolkit/lib/generate/name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const lodash_1 = tslib_1.__importDefault(require("lodash"));
/**
 * Stores a name with common transformations cached for use in templates and file paths.
 */
class Name {
    constructor(name) {
        this.original = "";
        this.kebabCase = "";
        this.lowerCamelCase = "";
        this.snakeCase = "";
        this.upperCamelCase = "";
        if (name) {
            this.original = name;
            this.kebabCase = lodash_1.default.kebabCase(name);
            this.lowerCamelCase = lodash_1.default.camelCase(name);
            this.snakeCase = lodash_1.default.snakeCase(name);
            this.upperCamelCase = lodash_1.default.upperFirst(lodash_1.default.camelCase(name));
        }
        Object.freeze(this);
    }
    toString() {
        return this.original;
    }
}
exports.Name = Name;
